binarySearch.js

```diff
@@ -0,0 +1,59 @@
+/*
+We're going to approach this problem using a binary search algorithm where we'll be able to search through
+a sorted list efficintly by repeatedly dividing a porition of the list in half, until we've narrow down the location
+of the target.
+
+We'll set two pointers, one at the start of the list and other at the end of the list.
+And from there, we'll do a comparison of values with the targert and depending
+on the outcome, we'll either move the left or right pointer until we're closer to the target.
+*/
+
+/*
+Question:
+Given an array of integers nums which is sorted in ascending order, and an integer target, write a function to search target in nums. 
+If target exists, then return its index. Otherwise, return -1.
+
+You must write an algorithm with O(log n) runtime complexity.
+*/
+const NOT_FOUND = -1;
+
+const getMidIx = (rightIdx, leftIdx) => {
+    const idxRange = rightIdx - leftIdx;
+    const midIdx = leftIdx + Math.floor(idxRange / 2);
+
+    return midIdx;
+};
+
+const getTargetIndex = (nums, target) => {
+    let leftIdx = 0;
+    let rightIdx = nums.length - 1;
+
+    while (leftIdx <= rightIdx) {
```

Comment:
P: What's the time complexity of this binary search and why?

Reply:
I'm still not that great with reading the time complexity, yet, but I believe it's O log(n) because instead of traversing through the entire input, we're cutting in half and re-adjusting the pointers, thus reducing the input size n.

Reply:
Yup, that's correct!

At each iteration, we look at our current section of the array, and we either say "the left half is invalid" or "the right half is invalid". So, we are cutting the array in half until we have a window that is very small (or until we just find the answer). In the worst case, we have to cut our array in half until the window size is 1.

The window size initially starts as `n`, the length of nums. Then, to know our time complexity, we ask the following question: how many times must I cut an array of length `n` in half until I get to a size 1 array? And the answer to this is `log n`.

Reply:
This is great. And just out of curiosity, is it O log(n) because since the size of n is decreasing, there's less computations? I'm assuming this based off a time complexity chart I googled.

Reply:
I'm not really sure what you mean by this.

It is O(log n) because we cut our array in half at each step.

For suppose, suppose our input is of length 32 (that is, n = 32). Then, we are claiming that we will have to do at most log 32 = 5 computations (this is technically not true because of some finer details of big O notation, but you don't need to worry about that). Also, note that 2^5 = 32 which is why log 32 = 5 (the base of our log is 2).

And this is correct! Here's an idea of how large our remaining array would be at each step:
32 elements (0 steps completed)
16 elements (1 step completed)
8 elements (2 steps completed)
4 elements (3 steps completed)
2 elements (4 steps completed)
1 element (5 steps completed)

We have solved our problem in 5 steps.